target/arc/fpu-helper.c

```diff
@@ -0,0 +1,826 @@
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see
+ * http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+
+#include "qemu/osdep.h"
+#include "cpu.h"
+#include "exec/exec-all.h"
+#include "exec/helper-proto.h"
+#include "fpu/softfloat.h"
+#include "fpu.h"
+#include "qemu/log.h"
+
+static uint8_t fpr_width;
+uint8_t vfp_width;
+/* TODO: Maybe is not needed */
+static uint8_t fpr_per_vector_operand;
+static uint8_t vfp_max_length;
+
+/*
+ * @brief Turn the internal soft fpu flags into an ARC FPU status value
+ */
+static uint32_t
+arc_pack_fpu_status(CPUARCState *env)
+{
+    uint32_t ret = 0;
+    uint8_t exp = get_float_exception_flags(&env->fp_status);
+
+    BIT_WRITE(ret, IV, READ_EXCP_FLAG(exp, float_flag_invalid));
+    BIT_WRITE(ret, DZ, READ_EXCP_FLAG(exp, float_flag_divbyzero));
+    BIT_WRITE(ret, OF, READ_EXCP_FLAG(exp, float_flag_overflow));
+    BIT_WRITE(ret, UF, READ_EXCP_FLAG(exp, float_flag_underflow));
+    BIT_WRITE(ret, IX, READ_EXCP_FLAG(exp, float_flag_inexact));
+    return ret;
+}
+
+/*
+ * @brief Turn the ARC FPU status value VAL into the internal soft fpu flags
+ */
+static void
+arc_unpack_fpu_status(CPUARCState *env, uint32_t val)
+{
+    uint32_t tmp = 0;
+    WRITE_EXCP_FLAG(tmp, float_flag_invalid,   BIT_READ(val, IV));
+    WRITE_EXCP_FLAG(tmp, float_flag_divbyzero, BIT_READ(val, DZ));
+    WRITE_EXCP_FLAG(tmp, float_flag_overflow,  BIT_READ(val, OF));
+    WRITE_EXCP_FLAG(tmp, float_flag_underflow, BIT_READ(val, UF));
+    WRITE_EXCP_FLAG(tmp, float_flag_inexact,   BIT_READ(val, IX));
+    set_float_exception_flags(tmp, &env->fp_status);
+}
+
+void
+init_fpu(CPUARCState *env, bool fp_dp, bool fp_wide, bool fp_hp)
+{
+    uint8_t index1 = ((fp_dp != 0) * 2) + (fp_wide != 0);
+    const uint8_t fpr_widths[] = { 32, 32, 64, 64 };
+    const uint8_t vfp_widths[] = { 32, 64, 64, 128 };
+
+    fpr_width = fpr_widths[index1];
+    vfp_width = vfp_widths[index1];
+
+    fpr_per_vector_operand = (vfp_width / fpr_width);
+    vfp_max_length = fpr_per_vector_operand *
+                     (vfp_width / 32) *
+                     (fp_hp != 0 ? 2 : 1);
+
+    uint32_t tmp = 0;
+    /*
+     * Disable default NaN mode (no platform default)
+     * NaN propagation is always done according to IEEE 754-2008
+     */
+    set_default_nan_mode(0, &env->fp_status);
+    /* As per ARCv3 PRM PU_CTRL Field Description */
+    set_float_rounding_mode(float_round_nearest_even, &env->fp_status);
+    /* Ensure exception flags start cleared */
+    set_float_exception_flags(tmp, &env->fp_status);
+
+    env->fp_status_persistent = arc_pack_fpu_status(env);
+}
+
+target_ulong
+arc_fpu_status_get_internal(CPUARCState *env)
```

Comment:
I don't understand this extra layer of internal-ness. everything could go inside "arc_fpu_status_set()".
Same holds true for `arc_fpu_status_get_internVal()`.